## 예술의전당 콘서트홀 가격 모델

### 1. 라이브러리

In [347]:
# base
import pandas as pd
import numpy as np
import re

# etc.
import sys
from glob import glob
import warnings
from tqdm import tqdm
import time 
import datetime as dt 
from scipy import stats


# visualization
import matplotlib.pyplot as plt
import koreanize_matplotlib
import plotly.express as px
import seaborn as sns
import matplotlib.colors as mcolors
from matplotlib.cm import ScalarMappable

# M.L
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.cluster import KMeans
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, silhouette_score
from kneed import KneeLocator

In [348]:
# settings
%matplotlib inline
pd.set_option("display.max_columns", 100)
pd.set_option('display.max_rows', 200)
warnings.filterwarnings(action = 'ignore')


### 2. 데이터셋 로드

In [349]:
# 기존의 제공된 데이터프레임에 빈좌석, 좌표, 무대와의 관계가 추가된 데이터프레임
file_name = "빈좌석_포함_클래식_데이터.parquet"
df = pd.read_parquet(glob(file_name)[0])

# 필요없는 컬럼 삭제
df = df.drop(['membership_type_1', 'membership_type_2', 'membership_type_3', 
              'membership_type_4', 'membership_type_5', 'membership_type_6',], axis=1)

print(f"""Found.. {len(glob(file_name))} file(s) : {glob(file_name)}
Reading.. {glob(file_name)[-1]}
df.shape : {df.shape}""")


Found.. 1 file(s) : ['빈좌석_포함_클래식_데이터.parquet']
Reading.. 빈좌석_포함_클래식_데이터.parquet
df.shape : (405810, 36)


In [350]:
df.loc[df['층'] == '합창석', 'Y'].describe()

count    44388.000000
mean      -793.423358
std        418.705553
min      -1295.000000
25%      -1195.000000
50%       -995.000000
75%       -410.000000
max         91.000000
Name: Y, dtype: float64

### 3. 클래스/함수 : 원가격, 원등급 추정

In [351]:
p_val = 2

class Performance:
    instance_cnt = 0

    
    def __init__(self, df):
        Performance.instance_cnt += 1
        self.instance_cnt = Performance.instance_cnt
        self.df = df
        self.perform_time = df['전체공연시간'].iloc[0]
        self.get_original_price()

    def get_original_price(self):
        """
        discount_type에서 할인율을 추출하고 역산해서 '할인전금액'을 df에 컬럼으로 추가하는 함수
        """
        # 할인이 적용되기 전 가격 추정
        self.df['할인율'] = self.df['discount_type'].str.extract('(\d+)%')
        self.df['할인율'] = self.df['할인율'].fillna(0).astype(int) / 100
        self.df['할인전가격'] = (self.df['price'].fillna(0) // (1 - self.df['할인율'])).round(-2).astype(int)
        self.df['할인전가격'] = self.df['할인전가격'].fillna(0)
        
        # 원가격추정 컬럼 추가 (5000원 단위로)
        self.df['원가격추정'] = ((self.df['할인전가격'] + 2500) // 5000 * 5000).copy()
        self.priced_seat = self.df[self.df['원가격추정'] > 0]
        self.unpriced_seat = self.df[self.df['원가격추정'] == 0]
        self.priced_rate = round((self.priced_seat.shape[0] / self.df.shape[0]), 3)
        self.booked_rate = round(self.df['예매여부'].mean(), 3)

    def get_best_n_neighbors(self):
        """
        knn 가격 추정 모델의 적절한 n_neighbors값을 찾는 함수
        """
        X = self.priced_seat.loc[self.priced_seat['층']!='합창석', ['X', 'Y', 'Z']] 
        y = self.priced_seat.loc[self.priced_seat['층']!='합창석', '원가격추정']
        cv_scores = []
        for n in range(1, min(50, (X.shape[0]*9//10)) + 1):
            model = KNeighborsRegressor(n_neighbors=n, weights='distance', p=p_val)
            scores = cross_val_score(model, X, y, cv=10, scoring='neg_mean_squared_error')
            cv_scores.append(scores.mean())
            self.best_n_neighbors = [i for i in range(1, 51)][np.argmax(cv_scores)]
            self.knn_mse = max(cv_scores) * -1

    def estimate_price(self):
        """
        knn모델에 추정된 n_neighbors값을 적용해서 판매되지 않은 티켓의 가격을 추정하는 함수
        """
        # 가격을 하나도 알 수 없는 경우 종료
        if self.priced_rate == 0:
            self.best_n_neighbors = 1
            self.knn_mse = 0
            self.mean_price = self.max_price = self.min_price = 0
            return
        
        # 1) 합창석이 아닌 좌석 : knn모델로 가젹 추정
        self.get_best_n_neighbors()  # 적절한 n_neighbors값 찾기
        X = self.priced_seat.loc[self.priced_seat['층']!='합창석', ['X', 'Y', 'Z']] 
        y = self.priced_seat.loc[self.priced_seat['층']!='합창석', '원가격추정']
        model = KNeighborsRegressor(n_neighbors=self.best_n_neighbors, weights='distance', p=p_val)
        model.fit(X, y)
        y_pred = model.predict(self.unpriced_seat.loc[self.unpriced_seat['층']!='합창석', ['X', 'Y', 'Z']])
        self.df.loc[(self.df['원가격추정']==0) & (self.df['층']!='합창석'), '원가격추정'] = y_pred
        
        # 2) 합창석 좌석 : 블록별 판매된 티켓의 최빈값으로 맵핑
        temp_dict = self.priced_seat.loc[self.priced_seat['층']=='합창석'].groupby('블록')['원가격추정'].agg(
            lambda x: x.mode().iloc[0] if not x.empty else None).to_dict()
        self.df.loc[self.df['층']=='합창석', '원가격추정'] = self.df.loc[self.df['층']=='합창석', '블록'].map(temp_dict)

        # 3) 가격 보정 : 이상치 보정, 5000원 단위로 표현
        threshold = 1.75  # 이상치 판단 기준
        self.df['원가격추정'] = self.df['원가격추정'].fillna(0)
        self.df['층블록'] = self.df['층'].astype(str) + self.df['블록'].astype(str)
        self.df['z_score'] = self.df.groupby('층블록')['원가격추정'].transform(lambda x: np.abs(stats.zscore(x))).round(2).fillna(0)
        temp_dict = self.df.groupby(['층', '블록'])['원가격추정'].median().to_dict()
        self.df.loc[self.df['z_score'] > threshold, '원가격추정'] = self.df.loc[self.df['z_score'] > threshold, '층블록'].map(temp_dict)
        self.df = self.df.drop('층블록', axis=1)
        self.df['원가격추정'] = ((self.df['원가격추정'].fillna(0) + 2500) // 5000 * 5000).astype(int)
 
    def estimate_cluster_kmeans(self):
        # 중복되는 값이 없도록 난수를 더해서 노이즈 만들기
        self.df['rand'] = np.random.rand(self.df.shape[0])
        self.df['원가격추정_rand'] = self.df['원가격추정'] + self.df['rand']

        # 군집화 모델 생성 (K-means)
        X = self.df[['원가격추정_rand']]
        inertia = []
        k_range = range(1, 11)
        for k in k_range:
            kmeans = KMeans(n_clusters=k, random_state=42, n_init='auto')
            kmeans.fit(X)
            inertia.append(kmeans.inertia_)
        
        # 적절한 K값 찾기 (elbow point)        
        kneedle = KneeLocator(k_range[1:], inertia[1:], curve='convex', direction='decreasing')
        self.best_k = kneedle.elbow

        # 찾은 K값을 적용해서 원등급 추정하기
        kmeans = KMeans(n_clusters=self.best_k, random_state=42, n_init='auto')
        self.df['원등급추정'] = kmeans.fit_predict(X)
        cluster_means = self.df.groupby('원등급추정').agg({'seat':'count', '원가격추정':'mean', '예매여부':'mean'}).reset_index()
        cluster_means = cluster_means.sort_values(by='원가격추정')
        grade_mapping = {grade: idx for idx, grade in enumerate(cluster_means['원등급추정'])}
        self.df = self.df.drop(['원가격추정_rand', 'rand'], axis=1)

        # 원등급, 등급별 가격의 비율, 군집분석의 실루엣점수
        self.df['원등급추정'] = (self.df['원등급추정'].map(grade_mapping) + 1)
        temp_min = min((x for x in cluster_means['원가격추정'] if x > 0), default=None)
        self.class_price_ratio = list((cluster_means['원가격추정'] / temp_min).round(2)) if temp_min else [0 for _ in range(len(cluster_means))]
        self.class_price = list(cluster_means['원가격추정'].round().astype(int))
        self.class_seats_cnt = list(cluster_means['seat'])
        self.class_booked_ratio = list(cluster_means['예매여부'].round(4))
        self.silhouette_score = silhouette_score(self.df[['원가격추정']], self.df['원등급추정'])
        self.mean_price = self.df['원가격추정'].mean().round(2)
        self.max_price = self.df['원가격추정'].max()
        self.min_price = self.df['원가격추정'].min()

    def px3dscatter(self, col_name):
        # 결과 시각화
        fig = px.scatter_3d(self.df, x='X', y='Y', z='Z', color=col_name, 
                            hover_name='seat', hover_data=['예매여부', '할인전가격', '원가격추정', 'z_score', '원등급추정'],
                            width=800, height=600)
        fig.update_traces(marker={'size': 1})
        fig.show()

### 4. 원가격/등급 추정 군집분석 결과

In [352]:
# 전체 데이터를 공연시간을 기준으로 공연별 분할
공연시간_list = sorted(df['전체공연시간'].unique())
공연별_df_list = [df[df['전체공연시간'] == 공연시간] for 공연시간 in 공연시간_list]

# 공연별 군집분석 결과 컬럼
instance_cnt_list = []
perform_time_list = []
priced_rate_list = []
booked_rate_list = []
best_n_neighbors_list = []
knn_mse_list = []
best_k_list = []
silhouette_score_list = []
mean_price_list = []
max_price_list = []
min_price_list = []

class_price_ratio_list = []
class_price_list = []
class_seats_cnt_list = []
class_booked_ratio_list = []
df_list = []

In [356]:
# class > 군집분석 적용
for 공연별_df in tqdm(공연별_df_list):
    p = Performance(공연별_df)
    p.estimate_price() # knn => 원가격 추정
    p.estimate_cluster_kmeans() # k-means => 원등급 추정

    instance_cnt_list.append(p.instance_cnt)
    perform_time_list.append(p.perform_time)
    priced_rate_list.append(p.priced_rate)
    booked_rate_list.append(p.booked_rate)
    best_n_neighbors_list.append(p.best_n_neighbors)
    knn_mse_list.append(p.knn_mse)
    best_k_list.append(p.best_k)
    silhouette_score_list.append(p.silhouette_score)
    mean_price_list.append(p.mean_price)
    max_price_list.append(p.max_price)
    min_price_list.append(p.min_price)
    class_price_ratio_list.append(p.class_price_ratio)
    class_price_list.append(p.class_price)
    class_seats_cnt_list.append(p.class_seats_cnt)
    class_booked_ratio_list.append(p.class_booked_ratio)
    df_list.append(p.df)

    # 시각화
    # print(p.perform_time)
    # p.px3dscatter('원등급추정')

# 군집결과 데이터프레임으로 합치기
df_result = pd.concat(df_list, axis=0)
df_summary = pd.DataFrame({
    '공연시간' : perform_time_list,
    '금액명시비율' : priced_rate_list,
    '예약율' : booked_rate_list,
    'knn_n_neighbors' : best_n_neighbors_list,
    'knn_mse' : knn_mse_list,
    'kmeans_군집수' : best_k_list,
    'kmeans_실루엣' : silhouette_score_list,
    '평균가격' : mean_price_list,
    '최소가격' : min_price_list,
    '최대가격' : max_price_list,
    '등급별가격비율' : class_price_ratio_list,
    '등급별가격' : class_price_list,
    '등급별좌석수' : class_seats_cnt_list,
    '등급별예매율' : class_booked_ratio_list
})

100%|██████████| 162/162 [04:42<00:00,  1.74s/it]


### 5. 결과 저장

* 총 162개의 클래식공연 중 11개의 공연은 가격이 전혀 명시되지 않았음.  => 해당 공연의 데이터에서는 원가격, 원등급 추정이 불가함.    
* 앞으로의 EDA를 위해 151개의 공연 데이터만 사용하기로 함.

#### 5-1) 공연별 요약 데이터 (df_summary)

In [357]:
# 공연별 (추정된) 가격, 등급 정보 요약 (군집분석 불가한 11개 공연 데이터는 제외)
공연별_원가격추정군집분석_결과요약 = df_summary.loc[df_summary['금액명시비율']!=0].reset_index().drop('index', axis=1)

# 컬럼 분리
공연별_원가격추정군집분석_결과요약['등급별가격_1'] = 공연별_원가격추정군집분석_결과요약['등급별가격비율'].map(lambda x: x[0])
공연별_원가격추정군집분석_결과요약['등급별가격_2'] = 공연별_원가격추정군집분석_결과요약['등급별가격비율'].map(lambda x: x[1])
공연별_원가격추정군집분석_결과요약['등급별가격_3'] = 공연별_원가격추정군집분석_결과요약['등급별가격비율'].map(lambda x: x[2] if len(x)>2 else np.nan)
공연별_원가격추정군집분석_결과요약['등급별가격_4'] = 공연별_원가격추정군집분석_결과요약['등급별가격비율'].map(lambda x: x[3] if len(x)>3 else np.nan)
공연별_원가격추정군집분석_결과요약['등급별가격_5'] = 공연별_원가격추정군집분석_결과요약['등급별가격비율'].map(lambda x: x[4] if len(x)>4 else np.nan)

# 데이터프레임 저장
공연별_원가격추정군집분석_결과요약.to_csv('공연별_원가격추정군집분석_결과요약.csv', index=False)
공연별_원가격추정군집분석_결과요약

공연시간  금액명시비율    예약율  knn_n_neighbors       knn_mse  \
0   2018-11-25 17:00:00   0.026  0.620                2  3.125890e+08   
1   2018-12-08 17:00:00   0.021  0.862                2  5.022582e+07   
2   2019-01-11 20:00:00   0.386  0.636                6  2.469253e+08   
3   2019-02-05 20:00:00   0.023  0.858                1  3.070000e+08   
4   2019-02-16 17:00:00   0.157  0.476               12  2.824275e+07   
5   2018-11-25 17:00:00   0.026  0.620                2  3.125890e+08   
6   2018-12-08 17:00:00   0.021  0.862                2  5.022582e+07   
7   2019-01-11 20:00:00   0.386  0.636                6  2.469253e+08   
8   2019-02-05 20:00:00   0.023  0.858                1  3.070000e+08   
9   2019-02-16 17:00:00   0.157  0.476               12  2.824275e+07   
10  2019-02-23 17:00:00   0.062  0.543                1  4.279167e+07   
11  2019-03-09 17:00:00   0.394  0.742                6  1.586679e+08   
12  2019-03-14 20:00:00   0.688  0.695               18  8.944243e+08   
13  2019-05-03 20:00:00   0.171  0.180               11  9.940577e+07   
14  2019-05-08 20:00:00   0.290  0.499               11  5.237612e+07   
15  2019-05-14 20:00:00   0.399  0.406                1  3.414931e+08   
16  2019-05-19 17:00:00   0.675  0.764                7  5.734333e+07   
17  2019-06-04 20:00:00   0.008  0.706                3  4.747094e+07   
18  2019-06-12 20:00:00   0.026  0.981                1  7.142857e+07   
19  2019-06-21 20:00:00   0.126  0.138               10  1.778024e+08   
20  2019-07-02 20:00:00   0.579  0.785                4  2.628508e+08   
21  2019-07-04 20:00:00   0.399  0.401               12  1.725214e+08   
22  2019-07-09 20:00:00   0.134  0.583               16  2.346811e+08   
23  2019-07-13 17:00:00   0.718  0.725               10  3.086576e+08   
24  2019-07-19 20:00:00   0.062  0.669                1  1.085125e+09   
25  2019-07-21 17:00:00   0.170  0.496                2  1.053229e+07   
26  2019-07-28 17:00:00   0.055  0.488                6  8.537000e+06   
27  2019-08-03 16:30:00   0.073  0.099                1  2.516082e+07   
28  2019-08-25 17:00:00   0.463  0.475                5  1.473965e+08   
29  2019-08-30 20:00:00   0.007  0.842                4  5.526210e+07   
30  2019-08-31 17:00:00   0.099  0.158                1  1.857296e-24   
31  2019-09-08 17:00:00   0.005  0.795               10  1.886743e+08   
32  2019-09-25 20:00:00   0.187  0.193                5  1.342522e+08   
33  2019-10-01 20:00:00   0.121  0.121               25  2.094853e+08   
34  2019-10-06 17:00:00   0.493  0.493                9  9.924147e+08   
35  2019-10-10 19:30:00   0.049  0.970                1  1.207051e+08   
36  2019-10-13 17:00:00   0.130  0.151                1  8.138731e+07   
37  2019-10-15 20:00:00   0.029  0.854                7  4.853037e+07   
38  2019-10-22 20:00:00   0.179  0.228                4  1.428689e+08   
39  2019-10-31 20:00:00   0.013  0.802               28  1.889880e+08   
40  2019-11-01 20:00:00   0.251  0.257               11  2.215497e+08   
41  2019-11-16 17:00:00   0.123  0.675                5  5.611620e+08   
42  2019-11-17 17:00:00   0.548  0.754                2  1.121103e+09   
43  2019-11-20 20:00:00   0.320  0.330               26  5.263388e+08   
44  2019-11-26 20:00:00   0.051  0.198               10  8.130186e+08   
45  2019-11-30 17:00:00   0.175  0.376                5  1.657986e+08   
46  2019-12-06 20:00:00   0.295  0.302                3  1.660594e+08   
47  2019-12-07 17:00:00   0.040  0.702                3  1.127337e+08   
48  2019-12-08 17:00:00   0.112  0.124               21  2.550634e+08   
49  2019-12-25 20:00:00   0.022  0.032                3  1.476346e+08   
50  2019-12-29 14:00:00   0.168  0.203                6  9.896237e+07   
51  2019-12-31 20:00:00   0.279  0.504               22  2.032730e+07   
52  2020-01-05 17:00:00   0.365  0.365               12  2.181251e+08   
53  2020-01-11 17:00:00   0.503  0.657                6  1.653156e+08   
54  

#### 5-2) 공연별 결과 데이터

In [358]:
# 전체공연좌석별 예매 데이터
공연별_결과_list = []
for 공연시간 in list(df_summary.loc[df_summary['금액명시비율']!=0, '공연시간']):
    공연별_결과_list.append(df_result[df_result['전체공연시간'] == 공연시간])

df_result = pd.concat(공연별_결과_list, axis=0, ignore_index=True)
print(f"""df_result 
memory usage: {round(df_result.memory_usage(deep=False).sum() / 1024**2, 1)}+ MB""")
df_result.to_parquet('전체공연_원가격원등급추정결과.parquet', index=False)

df_result 
memory usage: 122.2+ MB


In [359]:
df_result

seat   층   블록  열  넘버     X    Y    Z         대칭점      좌우시야각  \
0       1층 A블록1열 1  1층  A블록  1   1  1451  542  -93  1층 E블록1열 9  69.517555   
1       1층 A블록1열 2  1층  A블록  1   2  1406  555  -93  1층 E블록1열 8  68.459024   
2       1층 A블록1열 3  1층  A블록  1   3  1361  568  -93  1층 E블록1열 7  67.347261   
3       1층 A블록1열 4  1층  A블록  1   4  1315  580  -93  1층 E블록1열 6  66.199420   
4       1층 A블록1열 5  1층  A블록  1   5  1270  591  -93  1층 E블록1열 5  65.044882   
...            ...  ..  ... ..  ..   ...  ...  ...         ...        ...   
415825   2층 BOX6 2  2층  BOX  6   2 -1432  505  383   2층 BOX1 1 -70.574617   
415826   2층 BOX6 3  2층  BOX  6   3 -1500  565  428   2층 BOX1 4 -69.360281   
415827   2층 BOX6 4  2층  BOX  6   4 -1454  617  428   2층 BOX1 3 -67.006224   
415828   2층 BOX6 5  2층  BOX  6   5 -1522  677  483   2층 BOX1 6 -66.020049   
415829   2층 BOX6 6  2층  BOX  6   6 -1568  624  483   2층 BOX1 5 -68.299463   

            상하시야각    좌우면적시야각    상하면적시야각     무대까지의 거리  예매여부   age gender  \
0        3.436010  15.219900  33.948421  1551.713247     1   NaN    NaN   
1        3.520697  16.945594  34.331746  1514.433888     1   NaN    NaN   
2        3.608334  18.869853  34.695385  1477.698887     1   NaN    NaN   
3        3.702327  21.009736  35.069575  1440.234009     1   NaN    NaN   
4        3.798387  23.314833  35.426958  1403.862529     1   NaN    NaN   
...           ...        ...        ...          ...   ...   ...    ...   
415825 -14.156604  21.173142  34.169657  1565.994253     1  60.0      M   
415826 -14.950280  21.144470  32.190452  1659.038577     1  60.0      M   
415827 -15.161505  23.354990  31.913243  1636.456232     1  50.0      F   
415828 -16.169769  23.078793  30.097223  1734.388077     0   NaN    NaN   
415829 -15.971384  21.128253  30.390770  1755.360077     1   NaN    NaN   

        tran_date tran_time  play_date play_st_time    price  ticket_cancel  \
0      2018-11-16  15:06:00 2018-11-25     17:00:00      0.0            0.0   
1      2018-11-16  15:06:00 2018-11-25     17:00:00      0.0            0.0   
2      2018-11-16  15:06:00 2018-11-25     17:00:00      0.0            0.0   
3      2018-11-16  15:06:00 2018-11-25     17:00:00      0.0            0.0   
4      2018-11-16  15:06:00 2018-11-25     17:00:00      0.0            0.0   
...           ...       ...        ...          ...      ...            ...   
415825 2023-03-26  16:33:00 2023-06-06     19:30:00  15000.0            0.0   
415826 2023-03-26  16:33:00 2023-06-06     19:30:00  15000.0            0.0   
415827 2023-05-31  01:57:00 2023-06-06     19:30:00  28000.0            0.0   
415828        NaT      None 2023-06-06     19:30:00      NaN            0.0   
415829 2023-06-02  19:10:00 2023-06-06     19:30:00  30000.0            0.0   

          discount_type  performance_code pre_open_date  open_date  \
0                 기획사판매             715.0    2018-10-28 2018-10-29   
1                 기획사판매             715.0    2018-10-28 2018-10-29   
2                 기획사판매             715.0    2018-10-28 2018-10-29   
3                 기획사판매             715.0    2018-10-28 2018-10-29   
4                 기획사판매             715.0    2018-10-28 2018-10-29   
...                 ...               ...           ...        ...   
415825  장애인/국가유공자 할인50%            1508.0    2023-03-25 2023-03-26   
415826  장애인/국가유공자 할인50%            1508.0    2023-03-25 2023-03-26   
415827        그린회원 할인5%            1508.0    2023-03-25 2023-03-26   
415828              NaN            1508.0    2023-03-25 2023-03-26   
415829               일반            1508.0    2023-03-25 2023-03-26   

        running_time  intermission member_yn    공연연도   공연월         공연일  \
0              120.0          15.0         N  2018.0  11.0  2018-11-25   
1              120.0          15.0         N  2018.0  11.0  2018-11-25   
2              120.0          15.0         N  2018.0  11.0  2018-11-25   
3              120.0          15.0         N  2018.0  11.0  2018-11-25   
4              120.0        